README.md
# Ultima Loca

This generator enables (mixed !) narrow & wide rivers and well-defined mountain ranges with very flat areas inbetween.

It is still WIP but functional and safe to use.

The reason I made this was because I found the vanilla generator and other modded generators didn't create very scenic and practical maps because the transitions between highland and lowland was too gradual & 'bumpy' for my taste. Also, I wanted to have more control over rivers; especially their widths.

## Mountains
- Higher mountains (upto about 1500m) with a more distinct transition between flatland and cliffs.
- If you enable snow tops in the mod settings (on by default) mountain tops will be covered in snow above ~750m.
- NOTE: Mountain density is a fickle setting which seems to depend (internally) on the map seed. Esp. if you set the maximum mountain height to maximum. (I'm still finetuning this).

## Rivers
- Rivers can be much narrower and wider than in vanilla. (Keep in mind that ships cannot navigate very narrow rivers.)
- Wider rivers are also deeper (still experimenting with this).
- Lake generation in rivers can be disabled.
- More rivers.
- Curved rivers.
- Random river widths (still experimenting with this).
- Option to make rivers 'zig-zag' more (since v1.2).

## Rocks
- Option to disable scattered rocks on the map; completely or keep them along rivers.

## Trees & Forests
- Amount of scattered forests can be set (tree density).
- Tree density also affects the density of forests along mountain ridges.
- The treeline (height) on mountains can be changed.

### TODO:
- Option to add a coastline? (help needed)
- Scattered lakes (outside rivers)? (help needed)
- Translations
- Finetuning (mountain height-density ratio, river depth & width randomness)
- Known Issue: Rocks do not render when trees are set to 0.

### Important
If you enable the snow caps option, they will also appear in the (temperate) vanilla generator and any other modded generator that uses the temperate climate.


changelog.txt
1.2
- only change temperate climate for snow caps. Before desert/tropical were also affected.
- Added option for 'winding' rivers (zig-zag).
- Change default for mountain density to 'medium'
- Improved mountain-ridge tree densities.
- Improved snow layer
- Bugfix: error on random number when river is set to minimum
- Fixed: rocks didn't render when trees were set to 0. # of rocks is still lower (maxSlope is set for both trees & rocks in vanilla, I need to find a solution for this).

1.3
- Changed how mountain density is calculated (manual corrections added)
- Added the ability to change the mountain ridge curvature
- Removed minimum height of mountains (now set to 150 by default)
- Even wider rivers (this is the maximum before deformation occurs)
- More chance of branching rivers
- Changed some buttons to checkboxes (for UX and space)
- Changed name of mod


mod.lua
-- namespace = ulloc
local climate = require "ulloc/snowtops"

-- ------------------------------------------------------------------ --

function data()
  return {
    info = {
      name = "Not So Temperate Generator",
      description = "Temperate Terrain Generator that actually makes nice looking maps.",
      tags = { "Misc", "Script Mod" },
      authors = {
        {
          name = "Pentasis",
          role = 'CREATOR',
        },
      },
      minorVersion = 3,
      severityAdd = "NONE",
      severityRemove = "NONE",
      params = {
        {
          key = "ulloc_snowtops",
          name = "Snow caps on Mountains?",
          uiType = "BUTTON",
          values = { "Yes", "No" },
          defaultIndex = 0,
        },
      },
    },
    options = {},
    runFn = function(settings, modParams)
      local params = modParams[getCurrentModId()]
      if params.ulloc_snowtops == 0 then
        addModifier("loadClimate", climate.addSnowToTemperate)
      end
    end
    -- postRunFn = function (settings, params) end
  }
end


res/scripts/ulloc/snowtops.lua
local layersutil = require "terrain/layersutil"

local snowtops = {}

function snowtops.addSnowToTemperate(filename, climate)
  if climate ~= nil then
    if climate.updateFn ~= nil and climate.id == "temperate" then
      climate.updateFn = function (params)
        local result = {
          layers = layersutil.Layer.new(),
        }
        local heightmap = "heightmap"
        local mkTemp = layersutil.TempMaker.new()

        -- Greener grass
        local layer2_BaseRiverDistanceMin = 10
        local layer2_BaseRiverDistanceMax = 200 -- max: 256
        local layer2_NoiseStrength = 1.4
        local layer2_AmbientLevelsFrom = { 25, 150 } -- 0 to 255
        local layer2_AmbientLevelsTo = { 0.6, 1.0}
        -- Dark grass
        local layer4_AmbientLevelsFrom = { 100, 250 } -- 0 to 255
        local layer4_AmbientLevelsTo = { 0.9, 0 }
        -- Gravel beach
        local layerBeach_maxDistance = 80
        local layerBeach_gravelToStoneRatio = 0.4
        local layerBeach_gain = 0.9
        -- Cliffs
        local level2_CliffCutoff = { 0.7, 1.2 }

        -- #################
        -- #### PREPARE
        local distanceMap = mkTemp:Get()
        result.layers:Distance(heightmap, distanceMap, params.waterLevel)

        local ambientMap = mkTemp:Get()
        result.layers:AmbientOcclusion(heightmap, ambientMap, 14)

        -- #################
        -- #### Layer 2 - greenest grass
        local layer2Map = mkTemp:Get()
        result.layers:Map(distanceMap, layer2Map, { layer2_BaseRiverDistanceMin, layer2_BaseRiverDistanceMax }, { -1, 0}, true)

        local temp1Map = mkTemp:Get()
        result.layers:Map(ambientMap, temp1Map, layer2_AmbientLevelsFrom, layer2_AmbientLevelsTo, true)

        local noiseMap = mkTemp:Get()
        result.layers:PerlinNoise(noiseMap, {})
        result.layers:Map(noiseMap, noiseMap, {-1, 1}, {0, layer2_NoiseStrength}, true)

        local layer3Map = mkTemp:Get()
        result.layers:Mul(temp1Map, noiseMap, layer3Map)

        result.layers:Add(layer2Map, layer3Map, layer2Map)

        -- #################
        -- #### Layer 3 - Alpine grass
        do
          local temp2Map = mkTemp:Get()
          result.layers:Pwlerp(heightmap, temp2Map, { 300 , 400, 550, 620 } , { 0.5, 1, 0.5, 0 })
          result.layers:Mul(temp2Map, temp1Map, temp1Map)
          temp2Map = mkTemp:Restore(temp2Map)
        end

        result.layers:Mul(temp1Map, noiseMap, layer3Map)
        temp1Map = mkTemp:Restore(temp1Map)
        noiseMap = mkTemp:Restore(noiseMap)

        -- #################
        -- #### Layer 4 - Brown grass
        local layer4Map = mkTemp:Get()
        result.layers:Map(ambientMap, layer4Map, layer4_AmbientLevelsFrom, layer4_AmbientLevelsTo, false)

        do
          local temp1Map = mkTemp:Get()
          result.layers:Pwlerp(heightmap, temp1Map, { 500, 550, 600, 650, 700 }, { 0.5, 0.8, 1, 0.8, 0.5})
          result.layers:Mul(layer4Map, temp1Map, layer4Map)
          temp1Map = mkTemp:Restore(temp1Map)
        end

        -- Optional step
        result.layers:Pwlerp(layer4Map, layer4Map, {0, 0.1, 0.3, 1.0}, {0, 0, 1.0, 1.0})

        -- #################
        -- #### Layer 5 - Slopes
        local layer5Map, layer5bMap
        do
          local temp1Map = mkTemp:Get()
          result.layers:Map(distanceMap, temp1Map, { 150, 250 }, { 0.3, 1 }, true)

          result.layers:Mul(layer4Map, temp1Map, layer4Map)

          result.layers:Grad(heightmap, temp1Map, 3)

          layer5Map = mkTemp:Get()
          result.layers:Map(temp1Map, layer5Map, level2_CliffCutoff, { 0.0, 0.7 }, true)

          layer5bMap = mkTemp:Get()
          result.layers:Map(temp1Map, layer5bMap, level2_CliffCutoff, { 0.0, 0.8 }, true)
          temp1Map = mkTemp:Restore(temp1Map)
        end

        do
          local temp1Map = mkTemp:Get()
          result.layers:Laplace(heightmap, temp1Map)

          result.layers:Map(temp1Map, temp1Map, {0.2, 0.8 }, { 0.5, 0.7 }, true)

          result.layers:Mul(layer5bMap, temp1Map, layer5bMap)
          temp1Map = mkTemp:Restore(temp1Map)
        end

        -- #################
        -- #### Beach layer - gravel stone
        local layer6Map = mkTemp:Get()
        result.layers:Map(distanceMap, layer6Map, { 0, layerBeach_maxDistance }, { 2.3, 0}, true)
        distanceMap = mkTemp:Restore(distanceMap)

        do
          local noiseMap = mkTemp:Get()
          result.layers:RidgedNoise(noiseMap, { octaves = 3, lacunarity = 10.5, frequency = 1.0 / 1000.0, gain = layerBeach_gain})
          result.layers:Map(noiseMap, noiseMap, { 0.2, 0.8 }, { -2.3, 0}, true)
          result.layers:Add(layer6Map, noiseMap, layer6Map)
          mkTemp:Restore(noiseMap)
        end

        local layer7Map = mkTemp:Get()
        result.layers:Map(layer6Map, layer7Map, { 0.0, 1.0 }, { 0.0, layerBeach_gravelToStoneRatio}, true)

        -- #################
        -- #### Layer 8 - River bed
        local layer8Map = mkTemp:Get()
        result.layers:Map(heightmap, layer8Map, { -15.0, -1.0 }, { 1.0, 0.0}, true)

        -- #################
        -- #### Layer 9 - Snow
        local layer9Map = mkTemp:Get()
        result.layers:Map(heightmap, layer9Map, { 650 , 750 } , { 0.8, 1 }, true)
        ambientMap = mkTemp:Restore(ambientMap)

        do
          local temp1Map = mkTemp:Get()
          result.layers:Pwlerp(heightmap, temp1Map, { 600, 650, 750 }, { 0, 0.8, 1})
          result.layers:Mul(layer9Map, temp1Map, layer9Map)
          temp1Map = mkTemp:Restore(temp1Map)
        end
        -- #################
        -- #### Layer 10 - Brown under Snow
        local layer10Map = mkTemp:Get()
        result.layers:Map(heightmap, layer10Map, { 500 , 600 } , { 0.5, 1 }, true)

        do
          local temp1Map = mkTemp:Get()
          result.layers:Pwlerp(heightmap, temp1Map, { 400, 500, 600, 700 }, { 0, 0.5, 0.8, 1})
          result.layers:Mul(layer10Map, temp1Map, layer10Map)
          temp1Map = mkTemp:Restore(temp1Map)
        end

        -- #################
        -- #### Layer 11 - Mountain Grass
        local layer11Map = mkTemp:Get()
        result.layers:Map(heightmap, layer11Map, { 300 , 400, 500, 600 } , { 0.5, 1, 0.5, 0 }, true)

        do
          local temp1Map = mkTemp:Get()
          result.layers:Pwlerp(heightmap, temp1Map, { 250, 350, 450, 500, 550 }, { 0, 0.8, 1, 0.8, 0.3})
          result.layers:Mul(layer11Map, temp1Map, layer11Map)
          temp1Map = mkTemp:Restore(temp1Map)
        end

        result.mixingLayer = {
          backgroundMaterial = "grass_green.lua",
          layers = {
            -- Standard grass
            {
              map = layer2Map,
              dither = true,
              material = "grass_light_green.lua",
            },
            {
              map = layer4Map,
              dither = true,
              material = "scree.lua",
            },
            {
              map = layer11Map,
              dither = true,
              material = "grass_alpine.lua",
            },
            {
              map = layer3Map,
              dither = true,
              material = "grass_alpine.lua",
            },
            -- Snow scree materials
            {
              map = layer10Map,
              dither = true,
              material = "scree.lua",
            },
            -- Snow materials
            {
              map = layer9Map,
              dither = true,
              material = "snow_01.lua",
            },
            -- Slope rocks
            {
              map = layer5bMap,
              dither = true,
              material = "scree.lua",
            },
            {
              map = layer5Map,
              dither = true,
              material = "rock.lua",
            },
            -- Beach materials
            {
              map = layer6Map,
              dither = true,
              material = "gravel_01.lua",
            },
            {
              map = layer7Map,
              dither = true,
              material = "scree.lua",
            },
            -- River materials
            {
              map = layer8Map,
              dither = true,
              material = "scree.lua",
            },
          }
        }

        mkTemp:RestoreAll(result.mixingLayer)
        mkTemp:Finish()

        return result
      end
    end
  end

  return climate
end

return snowtops


res/scripts/ultima_loca/climate_snow.lua
local climate_snow = {}

function climate_snow.snowTops(filename, climate)
  if climate ~= nil then
    if climate.updateFn ~= nil then
      climate.updateFn = function (params)
        local result = {
          layers = layersutil.Layer.new(),
        }
        local heightmap = "heightmap"
        local mkTemp = layersutil.TempMaker.new()
        -- mkTemp.doDebug = true

        -- #################
        -- #### CONFIG
        -- Greener grass
        local layer2_BaseRiverDistanceMin = 10
        local layer2_BaseRiverDistanceMax = 200 -- max: 256
        local layer2_NoiseStrength = 1.4
        local layer2_AmbientLevelsFrom = { 15, 100 } -- 0 to 255
        local layer2_AmbientLevelsTo = { 0.7, 1.0}
        -- Dark grass
        -- Scree
        local layer4_AmbientLevelsFrom = { 200, 250 } -- 0 to 255
        local layer4_AmbientLevelsTo = { 0.9, 0.1 }
        -- Stone cliffs

        -- Gravel beach
        local layerBeach_maxDistance = 80
        local layerBeach_gravelToStoneRatio = 0.4
        local layerBeach_gain = 0.9
        -- Level 12: cliff
        local level2_CliffCutoff = { 0.7, 1.2 }

        -- #################
        -- #### PREPARE
        local distanceMap = mkTemp:Get()
        result.layers:Distance(heightmap, distanceMap, params.waterLevel)

        local ambientMap = mkTemp:Get()
        result.layers:AmbientOcclusion(heightmap, ambientMap, 14)

        -- #################
        -- #### Layer 2 - greenest grass

        local layer2Map = mkTemp:Get()
        result.layers:Map(distanceMap, layer2Map, { layer2_BaseRiverDistanceMin, layer2_BaseRiverDistanceMax }, { -1, 0}, true)

        local temp1Map = mkTemp:Get()
        result.layers:Map(ambientMap, temp1Map, layer2_AmbientLevelsFrom, layer2_AmbientLevelsTo, true)

        local noiseMap = mkTemp:Get()
        result.layers:PerlinNoise(noiseMap, {})
        result.layers:Map(noiseMap, noiseMap, {-1, 1}, {0, layer2_NoiseStrength}, true)

        local layer3Map = mkTemp:Get()
        result.layers:Mul(temp1Map, noiseMap, layer3Map)

        result.layers:Add(layer2Map, layer3Map, layer2Map)

        -- #################
        -- #### Layer 3 - Alpine grass
        do
          local temp2Map = mkTemp:Get()
          result.layers:Pwlerp(heightmap, temp2Map, { 300 , 400, 550, 620 } , { 0.5, 1, 0.5, 0 })
          result.layers:Mul(temp2Map, temp1Map, temp1Map)
          temp2Map = mkTemp:Restore(temp2Map)
        end

        result.layers:Mul(temp1Map, noiseMap, layer3Map)
        temp1Map = mkTemp:Restore(temp1Map)
        noiseMap = mkTemp:Restore(noiseMap)

        -- #################
        -- #### Layer 4 - Scree
        local layer4Map = mkTemp:Get()
        result.layers:Map(ambientMap, layer4Map, layer4_AmbientLevelsFrom, layer4_AmbientLevelsTo, false)


        do
          local temp1Map = mkTemp:Get()
          result.layers:Pwlerp(heightmap, temp1Map, { 500, 550, 600, 650, 700 }, { 0.5, 0.8, 1, 0.8, 0.5})
          result.layers:Mul(layer4Map, temp1Map, layer4Map)
          temp1Map = mkTemp:Restore(temp1Map)
        end

        -- Optional step
        result.layers:Pwlerp(layer4Map, layer4Map, {0, 0.1, 0.3, 1.0}, {0, 0, 1.0, 1.0})

        -- #################
        -- #### Layer 5 - Cliffs
        local layer5Map, layer5bMap
        do
          local temp1Map = mkTemp:Get()
          result.layers:Map(distanceMap, temp1Map, { 150, 250 }, { 0.3, 1 }, true)

          result.layers:Mul(layer4Map, temp1Map, layer4Map)

          result.layers:Grad(heightmap, temp1Map, 3)

          layer5Map = mkTemp:Get()
          result.layers:Map(temp1Map, layer5Map, level2_CliffCutoff, { 0.0, 0.7 }, true)

          layer5bMap = mkTemp:Get()
          result.layers:Map(temp1Map, layer5bMap, level2_CliffCutoff, { 0.0, 0.8 }, true)
          temp1Map = mkTemp:Restore(temp1Map)
        end

        do
          local temp1Map = mkTemp:Get()
          result.layers:Laplace(heightmap, temp1Map)

          result.layers:Map(temp1Map, temp1Map, {0.2, 0.8 }, { 0.5, 0.7 }, true)

          result.layers:Mul(layer5bMap, temp1Map, layer5bMap)
          temp1Map = mkTemp:Restore(temp1Map)
        end

        -- #################
        -- #### Layer 6, Layer 7 - Scree beach
        local layer6Map = mkTemp:Get()
        result.layers:Map(distanceMap, layer6Map, { 0, layerBeach_maxDistance }, { 2.3, 0}, true)
        distanceMap = mkTemp:Restore(distanceMap)

        do
          local noiseMap = mkTemp:Get()
          result.layers:RidgedNoise(noiseMap, { octaves = 3, lacunarity = 10.5, frequency = 1.0 / 1000.0, gain = layerBeach_gain})
          result.layers:Map(noiseMap, noiseMap, { 0.2, 0.8 }, { -2.3, 0}, true)
          result.layers:Add(layer6Map, noiseMap, layer6Map)
          mkTemp:Restore(noiseMap)
        end

        local layer7Map = mkTemp:Get()
        result.layers:Map(layer6Map, layer7Map, { 0.0, 1.0 }, { 0.0, layerBeach_gravelToStoneRatio}, true)

        -- #################
        -- #### Layer 8 - River bed
        local layer8Map = mkTemp:Get()
        result.layers:Map(heightmap, layer8Map, { -15.0, -1.0 }, { 1.0, 0.0}, true)

        -- #################
        -- #### Layer 9 - Snow caps
        local layer9Map = mkTemp:Get()
        result.layers:Map(heightmap, layer9Map, { 650 , 700 } , { 0.8, 1 }, true)
        ambientMap = mkTemp:Restore(ambientMap)

        do
          local temp1Map = mkTemp:Get()
          result.layers:Pwlerp(heightmap, temp1Map, { 620, 640, 700 }, { 0, 0.8, 1})
          result.layers:Mul(layer9Map, temp1Map, layer9Map)
          temp1Map = mkTemp:Restore(temp1Map)
        end
        -- #################
        -- #### Layer 10 - Scree under Snow
        local layer10Map = mkTemp:Get()
        result.layers:Map(heightmap, layer10Map, { 500 , 550 } , { 0.5, 1 }, true)

        do
          local temp1Map = mkTemp:Get()
          result.layers:Pwlerp(heightmap, temp1Map, { 400, 500, 550, 700 }, { 0, 0.5, 0.8, 1})
          result.layers:Mul(layer10Map, temp1Map, layer10Map)
          temp1Map = mkTemp:Restore(temp1Map)
        end

        -- #################
        -- #### Layer 11 - Alpine Grass
        local layer11Map = mkTemp:Get()
        result.layers:Map(heightmap, layer11Map, { 300 , 400, 550, 620 } , { 0.5, 1, 0.5, 0 }, true)

        do
          local temp1Map = mkTemp:Get()
          result.layers:Pwlerp(heightmap, temp1Map, { 250, 300, 400, 500, 520 }, { 0, 0.8, 1, 0.8, 0.3})
          result.layers:Mul(layer11Map, temp1Map, layer11Map)
          temp1Map = mkTemp:Restore(temp1Map)
        end

        -- #################
        -- #### MIX -- in sequence, so later materials are on top of earlier materials
        result.mixingLayer = {
          backgroundMaterial = "grass_green.lua",
          layers = {
            -- Standard grass
            {
              map = layer2Map,
              dither = true,
              material = "grass_light_green.lua",
            },
            {
              map = layer4Map,
              dither = true,
              material = "scree.lua",
            },
            {
              map = layer11Map,
              dither = true,
              material = "grass_alpine.lua",
            },
            {
              map = layer3Map,
              dither = true,
              material = "grass_alpine.lua",
            },

            -- Snow scree materials
            {
              map = layer10Map,
              dither = true,
              material = "scree.lua",
            },

            -- Snow materials
            {
              map = layer9Map,
              dither = true,
              material = "snow_01.lua",
            },

            -- Slope rocks
            {
              map = layer5bMap,
              dither = true,
              material = "scree.lua",
            },
            {
              map = layer5Map,
              dither = true,
              material = "rock.lua",
            },


            -- Beach materials
            {
              map = layer6Map,
              dither = true,
              material = "gravel_01.lua",
            },
            {
              map = layer7Map,
              dither = true,
              material = "scree.lua",
            },

            -- River materials
            {
              map = layer8Map,
              dither = true,
              material = "scree.lua",
            },


          }
        }

        mkTemp:RestoreAll(result.mixingLayer)
        mkTemp:Finish()

        --maputil.PrintGraph(result)
        return result
      end
    end
  end

  return climate
end


return climate_snow


res/config/terrain_generators/temperate_ulloc.gen.lua
local mapgenutil = require "terrain/ulloc_mapgenutil"
local temperateassetsgen = require "terrain/ulloc_temperateassetsgen"
local layersutil = require "terrain/layersutil"
local maputil = require "maputil"

function data()

  return {
    climate = "temperate.clima.lua",
    order = 0,
    name = _("Ultima Loca Temperate"),
    params = {
      --{
      --  key = "ulloc_mountain_low",
      --  name = _("Minimum Height Mountains"),
      --  values = { "", "", "", "", "", "", "", "", "", "" },
      --  defaultIndex = 5,
      --  uiType = "SLIDER",
      --  tooltip = "From 20 to 200m in steps of 20m.",
      --},
      {
        key = "ulloc_mountain_height",
        name = _("Mountain Height"),
        values = { "", "", "", "", "", "", "", "", "", "", "", "", "" },
        defaultIndex = 5,
        uiType = "SLIDER",
        tooltip = "From 300 to 1400m in steps of 100m.",
      },
      {
        key = "ulloc_mountain_seed",
        name = _("Mountain Range Curvature"),
        values = { "", "", "", "", "", "", "", "", "", "", "", "", "", "", "", "" },
        defaultIndex = 7,
        uiType = "SLIDER",
        tooltip = "Straight or curved mountain ranges.",
      },
      {
        key = "ulloc_terrain_ratio",
        name = _("Mountain Density"),
        values = { "Low", "Medium", "High" },
        defaultIndex = 1,
        uiType = "BUTTON",
      },
      {
        key = "ulloc_rivers",
        name = _("Amount of rivers"),
        values = { "", "", "", "", "", "", "", "", "", "", "" },
        defaultIndex = 5,
        uiType = "SLIDER",
      },
      {
        key = "ulloc_river_width",
        name = _("River width"),
        values = { "", "", "", "", "", "", "", "", "", "", "", "", "", "", "", "", "", "", "", "", "", "", "", "", "", "", "", "", "", "", "" },
        defaultIndex = 8,
        uiType = "SLIDER",
      },
      {
        key = "ulloc_curves",
        name = _("River curvature"),
        values = { "", "", "", "", "", "", "", "", "", "", "" },
        defaultIndex = 2,
        uiType = "SLIDER",
      },
      {
        key = "ulloc_rand_river",
        name = _("Random river widths"),
        values = { "0", "1" },
        defaultIndex = 0,
        uiType = "CHECKBOX",
      },
      {
        key = "ulloc_winding_river",
        name = _("Winding (zig-zag) rivers"),
        values = { "0", "1" },
        defaultIndex = 0,
        uiType = "CHECKBOX",
      },
      {
        key = "ulloc_river_lakes",
        name = _("Lakes in rivers"),
        values = { "0", "1" },
        defaultIndex = 1,
        uiType = "CHECKBOX",
      },
      --{
      --  key = "ulloc_lakes",
      --  name = _("Scattered Lakes?"),
      --  values = { "None", "Small", "Big", "Huge" },
      --  defaultIndex = 0,
      --  uiType = "COMBOBOX",
      --},
      {
        key = "ulloc_tree_density",
        name = _("Number of forests"),
        values = { "", "", "", "", "", "", "", "", "", "", "" },
        defaultIndex = 5,
        uiType = "SLIDER",
      },
      {
        key = "ulloc_treeline",
        name = _("Treeline height"),
        values = { "", "", "", "", "", "", "", "", "", "", "" },
        defaultIndex = 5,
        uiType = "SLIDER",
      },
      {
        key = "ulloc_rocks",
        name = _("Scattered Rocks?"),
        values = { "None", "Along rivers", "Everywhere" },
        defaultIndex = 1,
        uiType = "COMBOBOX",
      },
    },
    updateFn = function(params)
      print("seed: " .. math.random(1, 100000000))
      math.randomseed(math.random(1, 100000000))

      local result = {
        parallelFactor = 32,
        heightmapLayer = "HM",
        layers = layersutil.Layer.new(),
      }

      -- #################
      -- #### CONFIG
      local VALLEYS = {
        { 0, 1, 1, 1, 1, 2, 2, 2, 2, 3, 3, 3, 3 },
        { 0, 0, 0, 0, 0, 1, 1, 1, 1, 2, 2, 2, 2 },
        { 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1 },
      }
      local LOW_PROB = {
        { 0.4, 0.1 },
        { 0.5, 0.15 },
        { 0.6, 0.2 },
      }
      local HIGH_PROB = {
        { 0.1, 0 },
        { 0.15, 0 },
        { 0.2, 0 },
      }

      local mountain_height_selected = params.ulloc_mountain_height + 1
      local max_height_mountains = 200 + (100 * mountain_height_selected)
      local mountain_ratio = params.ulloc_terrain_ratio + 1

      local valley_amount = VALLEYS[mountain_ratio][mountain_height_selected]

      local delta_low_prob = (LOW_PROB[mountain_ratio][1] - LOW_PROB[mountain_ratio][2]) / 13
      local delta_high_prob = (HIGH_PROB[mountain_ratio][1] - HIGH_PROB[mountain_ratio][2]) / 13

      local mountain_low_probability = LOW_PROB[mountain_ratio][1] - (mountain_height_selected * delta_low_prob)
      local mountain_high_probability = HIGH_PROB[mountain_ratio][1] - (mountain_height_selected * delta_high_prob)

      local river_amount = params.ulloc_rivers / 10
      local humidity = params.ulloc_tree_density / 10
      local treeline = params.ulloc_treeline / 10

      local noWater = river_amount == 0

      local riverConfig = {
        depthScale = 1.5,
        maxOrder = river_amount * 10,
        segmentLength = 2400,
        bounds = params.bounds,
        baseProbability = river_amount * river_amount * 2,
        minDist = river_amount > 0.5 and 2 or 3,
        width = params.ulloc_river_width * 1.5,
        doRandomWidth = params.ulloc_rand_river == 1,
        curvature = params.ulloc_curves / 10,
        is_winding = params.ulloc_winding_river == 1,
      }

      local rivers = {}
      if not noWater then
        local start = mapgenutil.FindGoodRiverStart(params.bounds)
        mapgenutil.MakeRivers(rivers, riverConfig, 120000, start.pos, start.angle)

        if params.ulloc_river_lakes == 1 then
          local lakeConfig = {
            getLakePropability = function()
              return river_amount + 0.2
            end,
            lakeSize = river_amount * 600,
          }
          mapgenutil.MakeLakesOld(rivers, lakeConfig)
        end

        maputil.Convert(rivers)
        maputil.ValidateRiver(rivers)
      end

      local ridgesConfig = {
        bounds = params.bounds,
        probabilityLow = mountain_low_probability,
        probabilityHigh = mountain_high_probability,
        minHeight = 150,
        maxHeight = max_height_mountains,
        angle = params.ulloc_mountain_seed + 1
      }

      local valleys = {}
      if valley_amount > 0 then
        valleys = mapgenutil.MakeManyValleys(valley_amount)
      end
      local ridges = mapgenutil.MakeRidges(ridgesConfig)

      --local scatteredLakesConfig = {
      --  bounds = params.bounds,
      --  probabilityLow = 0.4,
      --  probabilityHigh = 0.0,
      --  minDepth = -20,
      --  maxDepth = -150,
      --}
--
      --local scatteredLakes = mapgenutil.MakeManyLakes(scatteredLakesConfig)

      -- #################
      -- #### PREPARE
      local mkTemp = layersutil.TempMaker.new()

      -- #################
      -- #### BACKGROUND AND RIVER
      result.layers:Constant(result.heightmapLayer, .01)

      do
        -- river
        result.layers:PushColor("#0022DD")
        result.layers:River(result.heightmapLayer, rivers)
        result.layers:PopColor()
      end

      local noiseMap = mkTemp:Get()
      result.layers:Noise(noiseMap, 0.1 * max_height_mountains)

      local distanceMap = mkTemp:Get()
      result.layers:Distance(result.heightmapLayer, distanceMap)

      -- #################
      -- #### RIDGES
      local ridgesMap
      do
        -- ridges
        result.layers:PushColor("#22AADD")
        local t1 = mkTemp:Get()
        result.layers:Map(distanceMap, t1, { 15, 2500 }, { 0, 1 }, true)
        result.layers:Mad(t1, noiseMap, result.heightmapLayer)
        noiseMap = mkTemp:Restore(noiseMap)

        local t2 = mkTemp:Get()
        result.layers:Ridge(t2, {
          valleys = valleys.points,
          ridges = ridges,
          noiseStrength = 10
        })

        result.layers:Map(distanceMap, t1, { 50, 1500 }, { 0, 1 }, true)
        ridgesMap = mkTemp:Get()
        result.layers:Mul(t1, t2, ridgesMap)
        t1 = mkTemp:Restore(t1)
        t2 = mkTemp:Restore(t2)

        result.layers:Add(ridgesMap, result.heightmapLayer, result.heightmapLayer)
        result.layers:PopColor()
      end

      -- #################
      -- #### Lakes
      --local scatteredLakesMap
      --do
      --  -- lakes
      --  result.layers:PushColor("#22AADD")
      --  local t1 = mkTemp:Get()
      --  result.layers:Map(distanceMap, t1, { 15, 2500 }, { 0, 1 }, true)
      --  result.layers:Mad(t1, noiseMap, result.heightmapLayer)
      --  noiseMap = mkTemp:Restore(noiseMap)
--
      --  local t2 = mkTemp:Get()
      --  result.layers:Ridge(t2, {
      --    valleys = {},
      --    ridges = scatteredLakes,
      --    noiseStrength = 10
      --  })
--
      --  result.layers:Map(distanceMap, t1, { 50, 1500 }, { 0, 1 }, true)
      --  scatteredLakesMap = mkTemp:Get()
      --  result.layers:Mul(t1, t2, scatteredLakesMap)
      --  t1 = mkTemp:Restore(t1)
      --  t2 = mkTemp:Restore(t2)
--
      --  result.layers:Add(scatteredLakesMap, result.heightmapLayer, result.heightmapLayer)
      --  result.layers:PopColor()
      --end

      -- #################
      -- #### NOISE
      local noiseStrength = 25.7
      local addNoise = true
      if addNoise then
        result.layers:PushColor("#5577DD")
        local t2 = mkTemp:Get()
        result.layers:RidgedNoise(t2, { octaves = 5, frequency = 1 / 444, lacunarity = 2.2, gain = 0.8 })
        result.layers:Map(t2, t2, { 0, 4 }, { 0, noiseStrength * 1.2 }, false)

        local t1 = mkTemp:Get()
        result.layers:Map(distanceMap, t1, { 0, 30 }, { 0, 1 }, true)

        result.layers:Mad(t2, t1, result.heightmapLayer)
        mkTemp:Restore(t1)
        mkTemp:Restore(t2)
        result.layers:PopColor()
      end

      -- #################
      -- #### ASSETS

      local config = {}

      config = {
        -- GENERIC
        humidity = humidity / 2.5,
        water = river_amount,
        -- LEVEL 3
        hillsLowLimit = 20, -- relative [m]
        hillsLowTransition = 20, -- relative [m]
        -- LEVEL 4
        treeLimit = treeline * ridgesConfig.maxHeight, --160, -- absolute [m] (absolute maximal height)
        ridgeFactor = 1 - (humidity / 10), -- lower means softer ridges detection, more trees (0.8)
        valleyFactor = 1 - (humidity / 10), -- lower means softer valleys detection, more trees (0.8)
        do_rocks = params.ulloc_rocks,
      }

      result.layers:PushColor("#007777")
      -- add scatteredLakesMap?
      result.forestMap, result.treesMapping, result.assetsMap, result.assetsMapping = temperateassetsgen.Make(
       result.layers, config, mkTemp, result.heightmapLayer, ridgesMap, distanceMap
      )
      result.layers:PopColor()
      distanceMap = nil

      -- #################
      -- #### FINISH
      ridgesMap = mkTemp:Restore(ridgesMap)
      --scatteredLakesMap = mkTemp:Restore(scatteredLakesMap)
      mkTemp:Restore(result.forestMap)
      mkTemp:Restore(result.assetsMap)
      mkTemp:Finish()
      -- maputil.PrintGraph(result)

      return result

    end
  }

end
